src/main/java/com/mitchej123/hodgepodge/SimulationDistanceHelper.java

```diff
@@ -0,0 +1,395 @@
+import net.minecraft.world.NextTickListEntry;
+import net.minecraft.world.World;
+import net.minecraft.world.WorldServer;
+import net.minecraftforge.common.ForgeChunkManager;
+
+import com.mitchej123.hodgepodge.config.FixesConfig;
+import com.mitchej123.hodgepodge.config.TweaksConfig;
+
+import it.unimi.dsi.fastutil.booleans.BooleanArrayList;
+import it.unimi.dsi.fastutil.longs.Long2BooleanOpenHashMap;
+import it.unimi.dsi.fastutil.longs.Long2ByteOpenHashMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.longs.LongArrayList;
+import it.unimi.dsi.fastutil.longs.LongIterator;
+import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
+
+public class SimulationDistanceHelper {
+
+    /**
+     * Mark a chunk as no to be simulated, or reset that state. Not thread safe!
+     */
+    public static void preventChunkSimulation(World world, long packedChunkPos, boolean prevent) {
+        if (!FixesConfig.addSimulationDistance) {
+            return;
+        }
+        ISimulationDistanceWorld mixin = (ISimulationDistanceWorld) world;
+        mixin.hodgepodge$preventChunkSimulation(packedChunkPos, prevent);
+    }
+
+    public static int getSimulationDistance() {
+        return TweaksConfig.simulationDistance;
+    }
+
+    public static void setSimulationDistance(int distance) {
+        TweaksConfig.simulationDistance = distance;
+    }
+
+    public SimulationDistanceHelper(World world) {
+        this.worldRef = new WeakReference<>(world);
+        isServer = world instanceof WorldServer;
+    }
+
+    /**
+     * The current world
+     */
+    private final WeakReference<World> worldRef;
+
+    /**
+     * True if we have a WorldServer
+     */
+    private final boolean isServer;
+
+    /**
+     * Chunks that should not be ticked
+     */
+    private final Long2ByteOpenHashMap noTickChunks = new Long2ByteOpenHashMap();
+
+    /**
+     * Changes to the chunks that should not be ticked. Needed since we only update the real map once a tick. Reason is
+     * that forced chunks are also cached and updated once a tick, we want both in sync.
+     */
+    private final LongBooleanArrayList noTickChunksChanges = new LongBooleanArrayList();
+
+    /**
+     * Cache of forced chunks. Profiler says that is faster.
+     */
+    private final LongOpenHashSet forcedChunksMap = new LongOpenHashSet();
+
+    /**
+     * Cache to quickly check if a chunk should be processed or not. Profiler says this is faster.
+     */
+    private final Long2BooleanOpenHashMap cacheShouldProcessTick = new Long2BooleanOpenHashMap();
+
+    /**
+     * Old player positions to see if they moved.
+     */
+    private final Map<EntityPlayer, ChunkCoordIntPair> playerPosOldMap = new WeakHashMap<>();
+
+    /**
+     * All ticks that are currently considered.
+     */
+    private final TreeSet<NextTickListEntry> pendingTickCandidates = new TreeSet<>();
```

Comment:
Might consider swapping this out with a fastutil treeset?

Reply:
Sure, done